django/contrib/admin/helpers.py

```diff
@@ -224,38 +224,44 @@ class InlineAdminFormSet(object):
     A wrapper around an inline formset for use in the admin system.
     """
     def __init__(self, inline, formset, fieldsets, prepopulated_fields=None,
-            readonly_fields=None, model_admin=None):
+            readonly_fields=None, uneditable_fields=None, has_add_permission=True,
+            model_admin=None):
         self.opts = inline
         self.formset = formset
         self.fieldsets = fieldsets
         self.model_admin = model_admin
         if readonly_fields is None:
             readonly_fields = ()
         self.readonly_fields = readonly_fields
+        if uneditable_fields is None:
+            uneditable_fields = ()
+        self.uneditable_fields = uneditable_fields
         if prepopulated_fields is None:
             prepopulated_fields = {}
         self.prepopulated_fields = prepopulated_fields
+        self.has_add_permission = has_add_permission
 
     def __iter__(self):
         for form, original in zip(self.formset.initial_forms, self.formset.get_queryset()):
             view_on_site_url = self.opts.get_view_on_site_url(original)
             yield InlineAdminForm(self.formset, form, self.fieldsets,
-                self.prepopulated_fields, original, self.readonly_fields,
+                self.prepopulated_fields, original, self.uneditable_fields,
                 model_admin=self.opts, view_on_site_url=view_on_site_url)
         for form in self.formset.extra_forms:
             yield InlineAdminForm(self.formset, form, self.fieldsets,
                 self.prepopulated_fields, None, self.readonly_fields,
                 model_admin=self.opts)
-        yield InlineAdminForm(self.formset, self.formset.empty_form,
-            self.fieldsets, self.prepopulated_fields, None,
-            self.readonly_fields, model_admin=self.opts)
+        if self.has_add_permission:
```

Comment:
Hi,

I'm trying to do the rebase on 1.11 in another PR (https://github.com/django/django/pull/6734), but I am having a little trouble understanding why you had to add this condition.

I don't see how adding the view permission requires to add a check on the add permission for the inlines.

With this change, I had some strange problems (the last object in the inline got hidden because it gets the empty-form class). Removing the condition leads to an exception which looks pretty low level, so I guess there was a good reason.

I could fix it (I think) with this commit :
https://github.com/olivierdalang/django/commit/061a2cbb98b1fafb52b14a2adc7c923fab5797bb

But I'm not sure of the fix at all, since I don't get why the change was made in the first place. 

Can you explain it to me ?

Thanks again for the great work !!

(besides, I see you have a 1.9.x backport branch for this, so do I, as I'll be needing this feature soon on 1.9.x, check it out if you're still working on this)